pkg/server/registry/apigroups/acorn/images/copy.go

```diff
@@ -0,0 +1,353 @@
+	client       kclient.WithWatch
+	transportOpt remote.Option
+}
+
+func (i *ImageCopy) NamespaceScoped() bool {
+	return true
+}
+
+func (i *ImageCopy) New() runtime.Object {
+	return &apiv1.ImageCopy{}
+}
+
+func (i *ImageCopy) NewConnectOptions() (runtime.Object, bool, string) {
+	return &apiv1.ImageCopy{}, false, ""
+}
+
+func (i *ImageCopy) Connect(ctx context.Context, _ string, _ runtime.Object, _ rest.Responder) (http.Handler, error) {
+	ns, _ := request.NamespaceFrom(ctx)
+
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		conn, err := k8schannel.Upgrader.Upgrade(rw, req, nil)
+		if err != nil {
+			logrus.Errorf("Error during handshake for image pull: %v", err)
+			return
+		}
+		defer conn.Close()
+
+		k8schannel.AddCloseHandler(conn)
+
+		args := &apiv1.ImageCopy{}
+		if err := conn.ReadJSON(args); err != nil {
+			_ = conn.CloseHandler()(websocket.CloseUnsupportedData, err.Error())
+			return
+		}
+
+		if err := i.Validate(ctx, *args, ns); err != nil {
+			_ = conn.CloseHandler()(websocket.CloseUnsupportedData, err.Error())
+			return
+		}
+
+		var progress <-chan ImageProgress
+		if !args.AllTags {
+			progress, err = i.ImageCopy(ctx, ns, args.Source, args.Dest, args.SourceAuth, args.DestAuth, args.Force)
+			if err != nil {
+				_ = conn.CloseHandler()(websocket.CloseInternalServerErr, err.Error())
+				return
+			}
+		} else {
+			progress, err = i.RepoCopy(ctx, ns, args.Source, args.Dest, args.SourceAuth, args.DestAuth, args.Force)
+			if err != nil {
+				_ = conn.CloseHandler()(websocket.CloseInternalServerErr, err.Error())
+				return
+			}
+		}
+
+		for update := range progress {
+			data, err := json.Marshal(update)
+			if err != nil {
+				panic("failed to marshal update: " + err.Error())
+			}
+			if err := conn.WriteMessage(websocket.TextMessage, data); err != nil {
+				logrus.Errorf("Error writing pull status: %v", err)
+				break
+			}
+		}
+
+		_ = conn.CloseHandler()(websocket.CloseNormalClosure, "")
+	}), nil
+}
+
+func (i *ImageCopy) ConnectMethods() []string {
+	return []string{"GET"}
+}
+
+func (i *ImageCopy) Validate(ctx context.Context, args apiv1.ImageCopy, namespace string) error {
+	if err := imagesystem.IsNotInternalRepo(ctx, i.client, namespace, args.Source); err != nil {
+		return err
+	}
+
+	if err := imagesystem.IsNotInternalRepo(ctx, i.client, namespace, args.Dest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *ImageCopy) ImageCopy(ctx context.Context, namespace, sourceImage, destImage string, srcAuth, dstAuth *apiv1.RegistryAuth, force bool) (<-chan ImageProgress, error) {
+	sourceRef, err := i.getImageReference(ctx, sourceImage, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if sourceRef.Context().RegistryStr() == images.NoDefaultRegistry {
+		return nil, fmt.Errorf("missing registry name (i.e. docker.io, ghcr.io) from source %s", sourceImage)
+	}
+
+	destRef, err := name.ParseReference(destImage, name.WithDefaultRegistry(images.NoDefaultRegistry))
```

Comment:
The destination is not allowed to be local, so we just parse the reference normally and ensure that there is an explicitly defined registry. Since `acorn push` doesn't implicitly use Docker Hub, I didn't want to implicitly use it here either, in order to be consistent.

Reply:
Same here.